### ISSUE TITLE
When I use db_paths ops, rocksdb write data to db_paths[0] & db_paths[3] path only

### ISSUE THREAD

**Issue by xiaobiaozhao** [facebook/rocksdb#10841]
rocks options
```
options_.db_paths = {{"/mnt/rocksdb/a", 1000 * 1000 * 1000},
                       {"/mnt/rocksdb/b", 1000 * 1000 * 1000},
                       {"/mnt/rocksdb/c", 1000 * 1000 * 1000},
                       {"/mnt/rocksdb/d", 1000 * 1000 * 1000}};
```
After put some data,   /mnt/rocksdb/a & /mnt/rocksdb/d will be writed sst file, but /mnt/rocksdb/a & /mnt/rocksdb/d is empty
```
du -h /mnt/rocksdb/
4.0K    /mnt/rocksdb/b
4.0K    /mnt/rocksdb/c
958M    /mnt/rocksdb/a
944M    /mnt/rocksdb/d
1.9G    /mnt/rocksdb/
```

### Expected behavior
According to the document description https://github.com/facebook/rocksdb/blob/main/include/rocksdb/options.h#L672
I think rocksdb will write sst file to  /mnt/rocksdb/a => /mnt/rocksdb/b => /mnt/rocksdb/c => /mnt/rocksdb/d

### Actual behavior
rocksdb write sst file  /mnt/rocksdb/a => /mnt/rocksdb/d，and /mnt/rocksdb/b & /mnt/rocksdb/c be skipped

### Steps to reproduce the behavior

```
// 1. set db_paths
options_.db_paths = {{"/mnt/rocksdb/a", 1000 * 1000 * 1000},
                       {"/mnt/rocksdb/b", 1000 * 1000 * 1000},
                       {"/mnt/rocksdb/c", 1000 * 1000 * 1000},
                       {"/mnt/rocksdb/d", 1000 * 1000 * 1000}};
// 2. opendb
rocksdb::DB::Open(options_, name, column_families, &cf_handles_, &db_);

// 3. put kv
db_->Put(rocksdb::WriteOptions(), key, value);
```

**Comment by mrambacher** [comment]
@xiaobiaozhao Is it possible that all of your data has been compacted into the lowest level?  Can you give an example of a complete test and execution showing this issue?  I have not had a chance to modify db_bench to show this problem.

**Comment by xiaobiaozhao** [comment]
> @xiaobiaozhao Is it possible that all of your data has been compacted into the lowest level? Can you give an example of a complete test and execution showing this issue? I have not had a chance to modify db_bench to show this problem.

Demo is here https://gist.github.com/xiaobiaozhao/75a0f6d3d3b3f564e28eacd9b85d3c1a

**Comment by aleksraiden** [comment]
Hi, any updates? Maybe in current 8.3.х this bug are fixed?

**Comment by xiaobiaozhao** [comment]
This is rocksdb v8.3.2
```
du -h /mnt/rocksdb/
471M    /mnt/rocksdb/a
4.0K    /mnt/rocksdb/c
1.5G    /mnt/rocksdb/d
4.0K    /mnt/rocksdb/b
```

**Comment by cxyxd** [comment]
please check the option:  level_compaction_dynamic_level_bytes
in rocksdb v8.3.2, the level_compaction_dynamic_level_bytes may be true

如果打开level_compaction_dynamic_level_bytes，则目标层会从默认的Level 1 变成最高层 Level 6，即最开始Level 0会直接compact到Level 6，如果某次compact后，Level 6大小超过256M(target_file_size_base)，假设300M，则base_level向上调整，此时base_level变成Level 5，而Level 5的大小上限是300M/10 = 30M，之后Level 0会直接compact到Level 5，如果Level 5超过30M，假设50M，则需要与Level 6进行compact，compact后，Level 5恢复到30M以下，Level 6稍微变大，假设320M，则基于320M继续调整base_level，即Level 5的大小上限，调整为320M/10 = 32M，随着写入持续进行，最终Level 5会超过256M(target_file_size_base)，此时base_level需要继续上调，到Level 4，取Level 5和Level 6当前大小较大者，记为MaxSize，则Level 4的大小上限为MaxSize/100，Level 5的大小上限为Level 4大小上限乘以10，依次类推。
相关代码在VersionStorageInfo::CalculateBaseBytes。

**Comment by xiaobiaozhao** [comment]
> @xiaobiaozhao
> 
> please check the option: level_compaction_dynamic_level_bytes in rocksdb v8.3.2, the level_compaction_dynamic_level_bytes may be true
> 
> 如果打开level_compaction_dynamic_level_bytes，则目标层会从默认的Level 1 变成最高层 Level 6，即最开始Level 0会直接compact到Level 6，如果某次compact后，Level 6大小超过256M(target_file_size_base)，假设300M，则base_level向上调整，此时base_level变成Level 5，而Level 5的大小上限是300M/10 = 30M，之后Level 0会直接compact到Level 5，如果Level 5超过30M，假设50M，则需要与Level 6进行compact，compact后，Level 5恢复到30M以下，Level 6稍微变大，假设320M，则基于320M继续调整base_level，即Level 5的大小上限，调整为320M/10 = 32M，随着写入持续进行，最终Level 5会超过256M(target_file_size_base)，此时base_level需要继续上调，到Level 4，取Level 5和Level 6当前大小较大者，记为MaxSize，则Level 4的大小上限为MaxSize/100，Level 5的大小上限为Level 4大小上限乘以10，依次类推。 相关代码在VersionStorageInfo::CalculateBaseBytes。

I'll test it when I have time.

**Comment by ptbxzrt** [comment]
> > [@xiaobiaozhao](https://github.com/xiaobiaozhao)
> > please check the option: level_compaction_dynamic_level_bytes in rocksdb v8.3.2, the level_compaction_dynamic_level_bytes may be true
> > 如果打开level_compaction_dynamic_level_bytes，则目标层会从默认的Level 1 变成最高层 Level 6，即最开始Level 0会直接compact到Level 6，如果某次compact后，Level 6大小超过256M(target_file_size_base)，假设300M，则base_level向上调整，此时base_level变成Level 5，而Level 5的大小上限是300M/10 = 30M，之后Level 0会直接compact到Level 5，如果Level 5超过30M，假设50M，则需要与Level 6进行compact，compact后，Level 5恢复到30M以下，Level 6稍微变大，假设320M，则基于320M继续调整base_level，即Level 5的大小上限，调整为320M/10 = 32M，随着写入持续进行，最终Level 5会超过256M(target_file_size_base)，此时base_level需要继续上调，到Level 4，取Level 5和Level 6当前大小较大者，记为MaxSize，则Level 4的大小上限为MaxSize/100，Level 5的大小上限为Level 4大小上限乘以10，依次类推。 相关代码在VersionStorageInfo::CalculateBaseBytes。
> 
> I'll test it when I have time.

hi, sorry to bother, any updates? Set level_compaction_dynamic_level_bytes to be false can fix this error?

**Comment by xiaobiaozhao** [comment]
> hi, sorry to bother, any updates? Set level_compaction_dynamic_level_bytes to be false can fix this error?

I did some testing and it didn't work out.

**Comment by ptbxzrt** [comment]
> > hi, sorry to bother, any updates? Set level_compaction_dynamic_level_bytes to be false can fix this error?
> 
> I did some testing and it didn't work out.

The reason you are observing data being written only to paths 'a' and 'd' is an expected result of RocksDB's path selection algorithm (`LevelCompactionBuilder::GetPathId()`).
Instead of simply choosing the path with the most free space, RocksDB performs a simulation to decide where to place files for each level. It predicts the total size of each level (L0, L1, L2, etc.) and tries to fit them sequentially into the `db_paths` you've defined.
Here's a breakdown of what happens with your configuration:
1. Path 'a' for L0/L1: The algorithm predicts the sizes for L0 and L1 (default: 256 MB each). Since they both fit within the 1 GB capacity of path 'a', they are assigned there.
2. L2 Prediction is Too Large: When it simulates placing L2, the predicted size is significantly larger (default: 2.56 GB, which is 256 MB * 10). This size does not fit in the remaining space of path 'a'.
3. Paths 'b' and 'c' are Skipped: The algorithm then checks the next paths. It sees that the predicted size of L2 (2.56 GB) is greater than the total capacity of path 'b' (1 GB) and path 'c' (1 GB). Because the algorithm determines these paths cannot even hold the single L2 level, it skips them entirely.
4. Path 'd' as Fallback: The last path in the list is always used as a fallback. Since L2 and all subsequent levels could not be placed in the preceding paths according to the simulation, they are all assigned to the final path, 'd'.
In short, paths 'b' and 'c' are skipped because their configured `target_size` is smaller than the predicted size of the first data level (L2) that needed a new home. To achieve a more even distribution, you would need to increase the `target_size` for each path to be larger than the predicted level sizes.

**Comment by xiaobiaozhao** [comment]
> > > hi, sorry to bother, any updates? Set level_compaction_dynamic_level_bytes to be false can fix this error?
> > 
> > 
> > I did some testing and it didn't work out.
> 
> The reason you are observing data being written only to paths 'a' and 'd' is an expected result of RocksDB's path selection algorithm (`LevelCompactionBuilder::GetPathId()`). Instead of simply choosing the path with the most free space, RocksDB performs a simulation to decide where to place files for each level. It predicts the total size of each level (L0, L1, L2, etc.) and tries to fit them sequentially into the `db_paths` you've defined. Here's a breakdown of what happens with your configuration:
> 
> 1. Path 'a' for L0/L1: The algorithm predicts the sizes for L0 and L1 (default: 256 MB each). Since they both fit within the 1 GB capacity of path 'a', they are assigned there.
> 2. L2 Prediction is Too Large: When it simulates placing L2, the predicted size is significantly larger (default: 2.56 GB, which is 256 MB * 10). This size does not fit in the remaining space of path 'a'.
> 3. Paths 'b' and 'c' are Skipped: The algorithm then checks the next paths. It sees that the predicted size of L2 (2.56 GB) is greater than the total capacity of path 'b' (1 GB) and path 'c' (1 GB). Because the algorithm determines these paths cannot even hold the single L2 level, it skips them entirely.
> 4. Path 'd' as Fallback: The last path in the list is always used as a fallback. Since L2 and all subsequent levels could not be placed in the preceding paths according to the simulation, they are all assigned to the final path, 'd'.
>    In short, paths 'b' and 'c' are skipped because their configured `target_size` is smaller than the predicted size of the first data level (L2) that needed a new home. To achieve a more even distribution, you would need to increase the `target_size` for each path to be larger than the predicted level sizes.

```c++
options_.db_paths = {{"/mnt/rocksdb/a", 1000 * 1000 * 1000},
                         {"/mnt/rocksdb/b", 10 * 1000 * 1000 * 1000L},
                         {"/mnt/rocksdb/c", 100 * 1000 * 1000 * 1000L},
                         {"/mnt/rocksdb/d", 1000 * 1000 * 1000 * 1000L}};
```
<img width="316" height="172" alt="Image" src="https://github.com/user-attachments/assets/ae6ab81a-10b3-4023-b3a0-2b5164ddb49c" />

Thank you, your explanation is correct. If I had four identical disks, they would do all the reading and writing for rocksdb. Do you have any suggestions for this need?
